{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/hbp5181/Linear-Model-uisng-homolog-survey-data/blob/main/future_learning(sequence_to_numbers).ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "43GQPw7R_ZbY",
      "metadata": {
        "id": "43GQPw7R_ZbY"
      },
      "source": [
        "# Inverse folding with ESM-IF1\n",
        "\n",
        "The ESM-IF1 inverse folding model is built for predicting protein sequences from their backbone atom coordinates. We provide examples here 1) to sample sequence designs for a given structure and 2) to score sequences for a given structure.\n",
        "\n",
        "Trained with 12M protein structures predicted by AlphaFold2, the ESM-IF1 model consists of invariant geometric input processing layers followed by a sequence-to-sequence transformer, and achieves 51% native sequence recovery on structurally held-out backbones. The model is also trained with span masking to tolerate missing backbone coordinates and therefore can predict sequences for partially masked structures.\n",
        "\n",
        "See [GitHub README](https://github.com/facebookresearch/esm/tree/main/examples/inverse_folding) for the complete user guide, and see our [bioRxiv pre-print](https://doi.org/10.1101/2022.04.10.487779) for more details."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "8c6TDUVupgNn",
      "metadata": {
        "id": "8c6TDUVupgNn"
      },
      "source": [
        "## Environment setup (colab)\n",
        "This step might take up to 10 minutes the first time.\n",
        "\n",
        "If using a local jupyter environment, instead of the following, we recommend configuring a conda environment upon first use in command line:\n",
        "```\n",
        "conda create -n inverse python=3.9\n",
        "conda activate inverse\n",
        "conda install pytorch cudatoolkit=11.3 -c pytorch\n",
        "conda install pyg -c pyg -c conda-forge\n",
        "conda install pip\n",
        "pip install biotite\n",
        "pip install git+https://github.com/facebookresearch/esm.git\n",
        "```\n",
        "\n",
        "Afterwards, `conda activate inverse` to activate this environment before starting `jupyter notebook`.\n",
        "\n",
        "Below is the setup for colab notebooks:\n",
        "\n",
        "We recommend using GPU runtimes on colab (Menu bar -> Runtime -> Change runtime type -> Hardware accelerator -> GPU)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "id": "nLOymXwdwUXo",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nLOymXwdwUXo",
        "outputId": "e598f2f5-9623-4645-ffd5-2e3a5fae9f44"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m10.8/10.8 MB\u001b[0m \u001b[31m36.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.0/5.0 MB\u001b[0m \u001b[31m13.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m10.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m932.1/932.1 kB\u001b[0m \u001b[31m4.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.0/1.0 MB\u001b[0m \u001b[31m7.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for fair-esm (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m52.8/52.8 MB\u001b[0m \u001b[31m10.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h"
          ]
        }
      ],
      "source": [
        "# Colab environment setup\n",
        "\n",
        "# Install the correct version of Pytorch Geometric.\n",
        "import torch\n",
        "import os\n",
        "\n",
        "def format_pytorch_version(version):\n",
        "  return version.split('+')[0]\n",
        "\n",
        "TORCH_version = torch.__version__\n",
        "TORCH = format_pytorch_version(TORCH_version)\n",
        "\n",
        "def format_cuda_version(version):\n",
        "  return 'cu' + version.replace('.', '')\n",
        "\n",
        "CUDA_version = torch.version.cuda\n",
        "CUDA = format_cuda_version(CUDA_version)\n",
        "\n",
        "!pip install -q torch-scatter -f https://data.pyg.org/whl/torch-{TORCH}+{CUDA}.html\n",
        "!pip install -q torch-sparse -f https://data.pyg.org/whl/torch-{TORCH}+{CUDA}.html\n",
        "!pip install -q torch-cluster -f https://data.pyg.org/whl/torch-{TORCH}+{CUDA}.html\n",
        "!pip install -q torch-spline-conv -f https://data.pyg.org/whl/torch-{TORCH}+{CUDA}.html\n",
        "!pip install -q torch-geometric\n",
        "\n",
        "# Install esm\n",
        "!pip install -q git+https://github.com/facebookresearch/esm.git\n",
        "\n",
        "# Install biotite\n",
        "!pip install -q biotite"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "EhDI4ZIX0Z4w",
      "metadata": {
        "id": "EhDI4ZIX0Z4w"
      },
      "source": [
        "### Verify that pytorch-geometric is correctly installed\n",
        "\n",
        "If the notebook crashes at the import, there is likely an issue with the version of torch_geometric and torch_sparse being incompatible with the torch version."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "id": "1-HvBXt1wWPu",
      "metadata": {
        "id": "1-HvBXt1wWPu"
      },
      "outputs": [],
      "source": [
        "## Verify that pytorch-geometric is correctly installed\n",
        "import torch_geometric\n",
        "import torch_sparse\n",
        "from torch_geometric.nn import MessagePassing"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "18544eee",
      "metadata": {
        "id": "18544eee"
      },
      "source": [
        "## Load model\n",
        "This steps takes a few minutes for the model to download.\n",
        "\n",
        "**UPDATE**: It is important to set the model in eval mode through `model = model.eval()` to disable random dropout for optimal performance."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "id": "14d8f393",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "14d8f393",
        "outputId": "76e2157d-ae64-4bcf-b200-8fd60ab54007"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Downloading: \"https://dl.fbaipublicfiles.com/fair-esm/models/esm_if1_gvp4_t16_142M_UR50.pt\" to /root/.cache/torch/hub/checkpoints/esm_if1_gvp4_t16_142M_UR50.pt\n",
            "/usr/local/lib/python3.10/dist-packages/esm/pretrained.py:215: UserWarning: Regression weights not found, predicting contacts will not produce correct results.\n",
            "  warnings.warn(\n"
          ]
        }
      ],
      "source": [
        "import esm\n",
        "model, alphabet = esm.pretrained.esm_if1_gvp4_t16_142M_UR50()\n",
        "model = model.eval()"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "17cb6adc",
      "metadata": {
        "id": "17cb6adc"
      },
      "source": [
        "## Extract encoder output as structure representation\n",
        "The encoder output may also be used as a representation for the structure.\n",
        "\n",
        "For a set of input coordinates with L amino acids, the encoder output will have shape L x 512."
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! esm-extract esm2_t33_650M_UR50D /content/RBD_aa_aligned.fasta\\\n",
        "  coordoutputRBD.fasta --repr_layers 33 --include mean\n",
        "! esm-extract esm2_t33_650M_UR50D /content/ACE2_aa_modified.fasta \\\n",
        "  coordoutputACE2.fasta --repr_layers 33 --include mean\n",
        "! esm-extract esm2_t33_650M_UR50D /content/ancestors_unique.fasta\\\n",
        "  coordoutputRBDa.fasta --repr_layers 33 --include mean\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zIV7yWTPG3Bc",
        "outputId": "37846f3f-f439-4ce6-f142-8b99d8e8aa71"
      },
      "id": "zIV7yWTPG3Bc",
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Transferred model to GPU\n",
            "Read /content/RBD_aa_aligned.fasta with 52 sequences\n",
            "Processing 1 of 3 batches (20 sequences)\n",
            "Processing 2 of 3 batches (20 sequences)\n",
            "Processing 3 of 3 batches (12 sequences)\n",
            "Transferred model to GPU\n",
            "Read /content/ACE2_aa_modified.fasta with 62 sequences\n",
            "Processing 1 of 13 batches (5 sequences)\n",
            "Processing 2 of 13 batches (5 sequences)\n",
            "Processing 3 of 13 batches (5 sequences)\n",
            "Processing 4 of 13 batches (5 sequences)\n",
            "Processing 5 of 13 batches (5 sequences)\n",
            "Processing 6 of 13 batches (5 sequences)\n",
            "Processing 7 of 13 batches (5 sequences)\n",
            "Processing 8 of 13 batches (5 sequences)\n",
            "Processing 9 of 13 batches (5 sequences)\n",
            "Processing 10 of 13 batches (5 sequences)\n",
            "Processing 11 of 13 batches (5 sequences)\n",
            "Processing 12 of 13 batches (5 sequences)\n",
            "Processing 13 of 13 batches (2 sequences)\n",
            "Transferred model to GPU\n",
            "Read /content/ancestors_unique.fasta with 34 sequences\n",
            "Processing 1 of 2 batches (20 sequences)\n",
            "Processing 2 of 2 batches (14 sequences)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Specify the folders containing the .pt files\n",
        "folder_paths = ['/content/coordoutputRBD.fasta', '/content/coordoutputACE2.fasta','/content/coordoutputRBDa.fasta']\n",
        "\n",
        "# Flatten the list of filenames\n",
        "pt_files = [os.path.join(folder, f) for folder in folder_paths for f in os.listdir(folder) if f.endswith('.pt')]\n",
        "\n",
        "# Iterate over each .pt file\n",
        "for file_path in pt_files:\n",
        "    # Load the model using torch.load\n",
        "    model_dict = torch.load(file_path, map_location=torch.device('cpu'))\n",
        "    for key, value in model_dict.items():\n",
        "        print(value)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aOd1C6URsuIs",
        "outputId": "864e55ec-0a8d-49e1-8558-e941f99ba841"
      },
      "id": "aOd1C6URsuIs",
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "RaTG13_MN996532\n",
            "{33: tensor([ 0.0063, -0.0141, -0.0844,  ..., -0.0330, -0.0599, -0.0377])}\n",
            "RmYN02_EPI_ISL_412977\n",
            "{33: tensor([ 0.0337, -0.0104, -0.0443,  ...,  0.0044, -0.0968, -0.0463])}\n",
            "Rs4084_KY417144\n",
            "{33: tensor([ 0.0156, -0.0255, -0.0592,  ..., -0.0306, -0.0504, -0.0517])}\n",
            "WIV1_KF367457\n",
            "{33: tensor([ 0.0231, -0.0028, -0.0825,  ..., -0.0678, -0.0402, -0.0350])}\n",
            "Rs4247_KY417148\n",
            "{33: tensor([ 0.0245, -0.0184, -0.0386,  ..., -0.0501, -0.0798, -0.0673])}\n",
            "Rf1_DQ412042\n",
            "{33: tensor([ 0.0473, -0.0279, -0.0498,  ...,  0.0203, -0.1102, -0.0508])}\n",
            "SARS-CoV-1_SZ13_PC03_AY304487\n",
            "{33: tensor([ 0.0095, -0.0220, -0.0646,  ..., -0.0386, -0.0439, -0.0491])}\n",
            "ZC45_MG772933\n",
            "{33: tensor([ 0.0231, -0.0156, -0.0568,  ...,  0.0006, -0.0960, -0.0460])}\n",
            "As6526_KY417142\n",
            "{33: tensor([ 0.0552, -0.0108, -0.0494,  ..., -0.0112, -0.0860, -0.0490])}\n",
            "SARS-CoV-2_MN908947\n",
            "{33: tensor([ 0.0224, -0.0117, -0.0591,  ..., -0.0322, -0.0758, -0.0261])}\n",
            "279-2005_DQ648857\n",
            "{33: tensor([ 0.0680, -0.0047, -0.0513,  ..., -0.0156, -0.0746, -0.0561])}\n",
            "JL2012_KJ473811\n",
            "{33: tensor([ 0.0557, -0.0231, -0.0534,  ...,  0.0089, -0.1066, -0.0383])}\n",
            "Hp-BCoV_Zhejiang_2013_KF636752\n",
            "{33: tensor([ 0.0736, -0.0184, -0.0271,  ..., -0.0709, -0.0943, -0.0486])}\n",
            "GX2013_KJ473815\n",
            "{33: tensor([ 0.0285, -0.0278, -0.0558,  ..., -0.0304, -0.0799, -0.0668])}\n",
            "SARS-CoV-1_SZ1_PC03_AY304489\n",
            "{33: tensor([ 0.0095, -0.0220, -0.0646,  ..., -0.0386, -0.0439, -0.0491])}\n",
            "ZXC21_MG772934\n",
            "{33: tensor([ 0.0150, -0.0062, -0.0693,  ..., -0.0126, -0.0851, -0.0503])}\n",
            "BM48-31_NC014470\n",
            "{33: tensor([ 0.0467, -0.0302, -0.0562,  ..., -0.0146, -0.0907, -0.0511])}\n",
            "SARS-CoV-1_Sino1-11_HP03L_AY485277\n",
            "{33: tensor([ 0.0071, -0.0249, -0.0671,  ..., -0.0364, -0.0440, -0.0448])}\n",
            "SARS-CoV-1_PC4-127_PC04_AY613951\n",
            "{33: tensor([ 0.0084, -0.0192, -0.0763,  ..., -0.0482, -0.0337, -0.0480])}\n",
            "SARS-CoV-1_GD03T0013_HP04_AY525636\n",
            "{33: tensor([ 0.0042, -0.0262, -0.0718,  ..., -0.0444, -0.0309, -0.0504])}\n",
            "Rf4092_KY417145\n",
            "{33: tensor([ 0.0147, -0.0205, -0.0313,  ...,  0.0061, -0.1000, -0.0434])}\n",
            "SARS-CoV-1_Urbani_HP03L_AY278741\n",
            "{33: tensor([ 0.0076, -0.0189, -0.0704,  ..., -0.0407, -0.0424, -0.0426])}\n",
            "Shaanxi2011_JX993987\n",
            "{33: tensor([ 0.0631, -0.0119, -0.0429,  ..., -0.0102, -0.0840, -0.0523])}\n",
            "SARS-CoV-1_GZ0402_HP04_AY613947\n",
            "{33: tensor([ 0.0026, -0.0196, -0.0872,  ..., -0.0545, -0.0335, -0.0362])}\n",
            "HKU3-13_GQ153548\n",
            "{33: tensor([ 0.0225, -0.0242, -0.0478,  ..., -0.0501, -0.0761, -0.0701])}\n",
            "Longquan-140_KF294457\n",
            "{33: tensor([ 0.0239, -0.0248, -0.0373,  ..., -0.0462, -0.0844, -0.0629])}\n",
            "HKU3-1_DQ022305\n",
            "{33: tensor([ 0.0225, -0.0242, -0.0478,  ..., -0.0501, -0.0761, -0.0701])}\n",
            "HuB2013_KJ473814\n",
            "{33: tensor([ 0.0285, -0.0190, -0.0411,  ..., -0.0357, -0.0779, -0.0696])}\n",
            "WIV16_KT444582\n",
            "{33: tensor([ 0.0231, -0.0028, -0.0825,  ..., -0.0678, -0.0402, -0.0350])}\n",
            "Rs7327_KY417151\n",
            "{33: tensor([ 0.0082, -0.0257, -0.0588,  ..., -0.0348, -0.0495, -0.0406])}\n",
            "SARS-CoV-1_HGZ8L1-A_HP03E_AY394981\n",
            "{33: tensor([ 0.0085, -0.0177, -0.0649,  ..., -0.0359, -0.0400, -0.0438])}\n",
            "Pangolin_GX-P2V_EPI_ISL_410542\n",
            "{33: tensor([ 0.0279, -0.0119, -0.0555,  ..., -0.0515, -0.0673, -0.0103])}\n",
            "SARS-CoV-1_GD01_HP03L_AY278489\n",
            "{33: tensor([ 0.0151, -0.0170, -0.0651,  ..., -0.0390, -0.0342, -0.0530])}\n",
            "BtKY72_KY352407\n",
            "{33: tensor([ 0.0296, -0.0303, -0.0700,  ..., -0.0195, -0.0567, -0.0671])}\n",
            "273-2005_DQ648856\n",
            "{33: tensor([ 0.0495, -0.0209, -0.0423,  ...,  0.0185, -0.1139, -0.0551])}\n",
            "RsSHC014_KC881005\n",
            "{33: tensor([ 0.0129, -0.0275, -0.0567,  ..., -0.0308, -0.0487, -0.0518])}\n",
            "Rp3_DQ071615\n",
            "{33: tensor([ 0.0500, -0.0083, -0.0523,  ..., -0.0240, -0.0853, -0.0465])}\n",
            "SARS-CoV-1_BJ02_HP03M_AY278487\n",
            "{33: tensor([ 0.0076, -0.0189, -0.0704,  ..., -0.0407, -0.0424, -0.0426])}\n",
            "HeB2013_KJ473812\n",
            "{33: tensor([ 0.0485, -0.0260, -0.0434,  ...,  0.0192, -0.1052, -0.0575])}\n",
            "Yunnan2011_JX993988\n",
            "{33: tensor([ 0.0345, -0.0024, -0.0536,  ...,  0.0030, -0.0974, -0.0406])}\n",
            "Rs4237_KY417147\n",
            "{33: tensor([ 0.0519, -0.0114, -0.0525,  ..., -0.0140, -0.0871, -0.0481])}\n",
            "SARS-CoV-1_Sin852_HP03L_AY559082\n",
            "{33: tensor([ 0.0048, -0.0156, -0.0731,  ..., -0.0422, -0.0449, -0.0383])}\n",
            "SARS-CoV-1_PC4-137_PC04_AY627045\n",
            "{33: tensor([ 0.0064, -0.0248, -0.0798,  ..., -0.0482, -0.0354, -0.0502])}\n",
            "YN2013_KJ473816\n",
            "{33: tensor([ 0.0143, -0.0041, -0.0727,  ...,  0.0055, -0.0838, -0.0385])}\n",
            "LYRa11_KF569996\n",
            "{33: tensor([ 0.0084, -0.0171, -0.0563,  ..., -0.0425, -0.0463, -0.0373])}\n",
            "HKU3-8_GQ153543\n",
            "{33: tensor([ 0.0299, -0.0198, -0.0261,  ..., -0.0393, -0.0915, -0.0591])}\n",
            "Rs4081_KY417143\n",
            "{33: tensor([ 0.0448, -0.0088, -0.0608,  ..., -0.0192, -0.0800, -0.0466])}\n",
            "Rs4231_KY417146\n",
            "{33: tensor([ 0.0147, -0.0262, -0.0568,  ..., -0.0316, -0.0506, -0.0525])}\n",
            "Pangolin_GD-consensus_Lam2020\n",
            "{33: tensor([ 0.0250, -0.0273, -0.0550,  ..., -0.0276, -0.0702, -0.0375])}\n",
            "SARS-CoV-1_SZ3_PC03_AY304486\n",
            "{33: tensor([ 0.0095, -0.0220, -0.0646,  ..., -0.0386, -0.0439, -0.0491])}\n",
            "SARS-CoV-1_PC4-13_PC04_AY613948\n",
            "{33: tensor([ 0.0090, -0.0179, -0.0779,  ..., -0.0510, -0.0384, -0.0454])}\n",
            "SARS-CoV-1_GZ-C_HP03L_AY394979\n",
            "{33: tensor([-0.0008, -0.0186, -0.0737,  ..., -0.0498, -0.0415, -0.0374])}\n",
            "R_affinis_9488\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_affinis_9497\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "Pangolin_ACE2_XP017505746\n",
            "{33: tensor([ 0.0477, -0.0602, -0.0599,  ..., -0.1004, -0.0132,  0.0648])}\n",
            "R_affinis_9498\n",
            "{33: tensor([ 0.0463, -0.0616, -0.0435,  ..., -0.1001, -0.0084,  0.0622])}\n",
            "Mouse_ACE2_NP001123985\n",
            "{33: tensor([ 0.0552, -0.0543, -0.0560,  ..., -0.1079, -0.0183,  0.0739])}\n",
            "R_sinicus_3366_Yunnan_allele4\n",
            "{33: tensor([ 0.0489, -0.0633, -0.0492,  ..., -0.0973, -0.0094,  0.0667])}\n",
            "R_sinicus_411_Guangxi_allele8_GQ999933\n",
            "{33: tensor([ 0.0453, -0.0653, -0.0539,  ..., -0.0996, -0.0097,  0.0624])}\n",
            "R_affinis_9500\n",
            "{33: tensor([ 0.0463, -0.0616, -0.0435,  ..., -0.1001, -0.0084,  0.0622])}\n",
            "Human_ACE2_NP001358344\n",
            "{33: tensor([ 0.0508, -0.0604, -0.0513,  ..., -0.0982, -0.0180,  0.0772])}\n",
            "R_affinis_5538\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_sinicus_WJ1_Hubei_allele5\n",
            "{33: tensor([ 0.0456, -0.0631, -0.0500,  ..., -0.1003, -0.0060,  0.0638])}\n",
            "R_sinicus_1443_Guangdong_allele4\n",
            "{33: tensor([ 0.0465, -0.0635, -0.0489,  ..., -0.0997, -0.0070,  0.0644])}\n",
            "R_affinis_9493\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_affinis_9496\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_affinis_9495\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "Mustela_lutreola_ACE2_QNC68911\n",
            "{33: tensor([ 0.0513, -0.0434, -0.0499,  ..., -0.1078, -0.0149,  0.0682])}\n",
            "R_sinicus_984_Hubei_allele5\n",
            "{33: tensor([ 0.0445, -0.0627, -0.0498,  ..., -0.1010, -0.0058,  0.0638])}\n",
            "R_affinis_787\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_sinicus_1450_Guangdong_allele8\n",
            "{33: tensor([ 0.0452, -0.0646, -0.0517,  ..., -0.0978, -0.0081,  0.0627])}\n",
            "RaccoonDog_ACE2_ABW16956\n",
            "{33: tensor([ 0.0590, -0.0518, -0.0481,  ..., -0.1041, -0.0175,  0.0613])}\n",
            "R_sinicus_5718_Yunnan_allele1\n",
            "{33: tensor([ 0.0473, -0.0627, -0.0496,  ..., -0.0979, -0.0080,  0.0650])}\n",
            "R_affinis_9482\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_affinis_9492\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_affinis_9508\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_affinis_9470\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "Neovison_vison_ACE2_MW269526\n",
            "{33: tensor([ 0.0498, -0.0434, -0.0536,  ..., -0.1071, -0.0144,  0.0646])}\n",
            "R_sinicus_1438_Guangdong_allele8\n",
            "{33: tensor([ 0.0447, -0.0649, -0.0522,  ..., -0.0999, -0.0067,  0.0629])}\n",
            "R_sinicus_WJ6_Hubei_allele5\n",
            "{33: tensor([ 0.0445, -0.0627, -0.0498,  ..., -0.1010, -0.0058,  0.0638])}\n",
            "Civet_ACE2_Q56NL1\n",
            "{33: tensor([ 0.0513, -0.0555, -0.0516,  ..., -0.1029, -0.0176,  0.0775])}\n",
            "R_sinicus_1448_Guangdong_allele4\n",
            "{33: tensor([ 0.0463, -0.0631, -0.0497,  ..., -0.1003, -0.0068,  0.0638])}\n",
            "R_sinicus_1446_Guangdong_allele4\n",
            "{33: tensor([ 0.0465, -0.0635, -0.0489,  ..., -0.0997, -0.0070,  0.0644])}\n",
            "R_landeri_ACE2_ALJ94034\n",
            "{33: tensor([ 0.0414, -0.0630, -0.0506,  ..., -0.0993, -0.0146,  0.0643])}\n",
            "R_affinis_9501\n",
            "{33: tensor([ 0.0468, -0.0624, -0.0424,  ..., -0.0999, -0.0099,  0.0625])}\n",
            "R_affinis_9486\n",
            "{33: tensor([ 0.0449, -0.0600, -0.0450,  ..., -0.1031, -0.0077,  0.0596])}\n",
            "R_affinis_9489\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_sinicus_1433_Guangdong_allele4\n",
            "{33: tensor([ 0.0457, -0.0636, -0.0497,  ..., -0.1011, -0.0063,  0.0643])}\n",
            "Neovison_vison_ACE2partial_HAAF01014901\n",
            "{33: tensor([ 0.0781, -0.0399, -0.0837,  ..., -0.0827, -0.0534,  0.0505])}\n",
            "R_affinis_9476\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_sinicus_WJ2_Hubei_allele5\n",
            "{33: tensor([ 0.0445, -0.0627, -0.0498,  ..., -0.1010, -0.0058,  0.0638])}\n",
            "R_sinicus_5720_Yunnan_allele2\n",
            "{33: tensor([ 0.0475, -0.0648, -0.0496,  ..., -0.1010, -0.0080,  0.0629])}\n",
            "R_sinicus_3364_Yunnan_allele1\n",
            "{33: tensor([ 0.0468, -0.0627, -0.0500,  ..., -0.0973, -0.0077,  0.0655])}\n",
            "R_alcyone_ACE2_ALJ94035\n",
            "{33: tensor([ 0.0425, -0.0623, -0.0521,  ..., -0.0994, -0.0133,  0.0624])}\n",
            "R_sinicus_3358_Yunnan_allele7\n",
            "{33: tensor([ 0.0478, -0.0655, -0.0510,  ..., -0.0970, -0.0094,  0.0669])}\n",
            "R_ferrumequinum_ACE2_ADN93470_Houetal\n",
            "{33: tensor([ 0.0471, -0.0660, -0.0494,  ..., -0.0992, -0.0143,  0.0600])}\n",
            "R_sinicus_3359_Yunnan_allele1\n",
            "{33: tensor([ 0.0476, -0.0629, -0.0498,  ..., -0.0961, -0.0085,  0.0665])}\n",
            "R_sinicus_1455_Guangdong_allele4\n",
            "{33: tensor([ 0.0465, -0.0635, -0.0489,  ..., -0.0997, -0.0070,  0.0644])}\n",
            "R_macrotis_ACE2_ADN93471\n",
            "{33: tensor([ 0.0444, -0.0636, -0.0497,  ..., -0.1022, -0.0077,  0.0602])}\n",
            "R_pearsonii_ACE2_ABU54053\n",
            "{33: tensor([ 0.0370, -0.0635, -0.0494,  ..., -0.0993, -0.0053,  0.0632])}\n",
            "R_sinicus_3357_Yunnan_allele7_KC881004\n",
            "{33: tensor([ 0.0480, -0.0657, -0.0519,  ..., -0.0971, -0.0094,  0.0660])}\n",
            "R_affinis_9479\n",
            "{33: tensor([ 0.0431, -0.0607, -0.0453,  ..., -0.1014, -0.0063,  0.0595])}\n",
            "R_affinis_9490\n",
            "{33: tensor([ 0.0458, -0.0611, -0.0437,  ..., -0.1005, -0.0076,  0.0584])}\n",
            "R_affinis_9491\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_sinicus_WJ4_Hubei_allele2\n",
            "{33: tensor([ 0.0475, -0.0648, -0.0496,  ..., -0.1010, -0.0080,  0.0629])}\n",
            "R_affinis_9473\n",
            "{33: tensor([ 0.0467, -0.0617, -0.0435,  ..., -0.0993, -0.0088,  0.0612])}\n",
            "R_sinicus_1449_Guangdong_allele8\n",
            "{33: tensor([ 0.0447, -0.0649, -0.0522,  ..., -0.0999, -0.0067,  0.0629])}\n",
            "R_sinicus_ACT66275.1_HongKong_allele3_GQ262791\n",
            "{33: tensor([ 0.0481, -0.0648, -0.0483,  ..., -0.1020, -0.0081,  0.0639])}\n",
            "R_sinicus_1434_Guangdong_allele6\n",
            "{33: tensor([ 0.0464, -0.0636, -0.0489,  ..., -0.1000, -0.0072,  0.0653])}\n",
            "R_pusillus_ACE2_ADN93477\n",
            "{33: tensor([ 0.0467, -0.0594, -0.0460,  ..., -0.1059, -0.0071,  0.0662])}\n",
            "R_sinicus_3362_Yunnan_allele1\n",
            "{33: tensor([ 0.0476, -0.0629, -0.0498,  ..., -0.0961, -0.0085,  0.0665])}\n",
            "R_affinis_9494\n",
            "{33: tensor([ 0.0464, -0.0621, -0.0433,  ..., -0.0984, -0.0102,  0.0612])}\n",
            "R_sinicus_832_Hubei_allele4_GQ999936\n",
            "{33: tensor([ 0.0440, -0.0634, -0.0507,  ..., -0.1025, -0.0060,  0.0613])}\n",
            "R_affinis_4331\n",
            "{33: tensor([ 0.0449, -0.0600, -0.0450,  ..., -0.1031, -0.0077,  0.0596])}\n",
            "AncEurAf_tree1\n",
            "{33: tensor([ 0.0409, -0.0356, -0.0473,  ..., -0.0226, -0.0616, -0.0700])}\n",
            "AncSarbecovirus_tree1\n",
            "{33: tensor([ 0.0388, -0.0354, -0.0500,  ..., -0.0245, -0.0616, -0.0686])}\n",
            "AncClade2_alt2_del1-only\n",
            "{33: tensor([ 0.0429, -0.0256, -0.0640,  ..., -0.0326, -0.0553, -0.0394])}\n",
            "AncSARS1a_tree2\n",
            "{33: tensor([-0.0232, -0.0553, -0.0567,  ..., -0.0216, -0.0627, -0.1054])}\n",
            "AncSARS2a_MAP\n",
            "{33: tensor([ 0.0359, -0.0191, -0.0611,  ..., -0.0384, -0.0611, -0.0403])}\n",
            "AncClade2_alt3_del2-only\n",
            "{33: tensor([ 0.0325, -0.0057, -0.0603,  ..., -0.0508, -0.0531, -0.0547])}\n",
            "AncSARS1b_MAP\n",
            "{33: tensor([ 0.0082, -0.0257, -0.0588,  ..., -0.0348, -0.0495, -0.0406])}\n",
            "AncSARS1a_altALL\n",
            "{33: tensor([ 0.0504, -0.0259, -0.0451,  ..., -0.0230, -0.0652, -0.0641])}\n",
            "AncClade2_altALL\n",
            "{33: tensor([ 0.0356, -0.0335, -0.0593,  ..., -0.0217, -0.0786, -0.0750])}\n",
            "AncSARS-CoV-1_MAP\n",
            "{33: tensor([ 0.0106, -0.0189, -0.0610,  ..., -0.0418, -0.0390, -0.0469])}\n",
            "AncSARS-CoV-1_04_MAP\n",
            "{33: tensor([ 0.0064, -0.0248, -0.0798,  ..., -0.0482, -0.0354, -0.0502])}\n",
            "AncAsia_tree2\n",
            "{33: tensor([ 0.0377, -0.0165, -0.0584,  ..., -0.0393, -0.0592, -0.0347])}\n",
            "AncSARS2b_MAP\n",
            "{33: tensor([ 0.0261, -0.0281, -0.0530,  ..., -0.0307, -0.0705, -0.0272])}\n",
            "AncSARS-CoV-1_04_human_MAP\n",
            "{33: tensor([ 0.0042, -0.0262, -0.0718,  ..., -0.0444, -0.0309, -0.0504])}\n",
            "AncAsia_MAP\n",
            "{33: tensor([ 0.0361, -0.0178, -0.0582,  ..., -0.0418, -0.0524, -0.0402])}\n",
            "AncClade2_alt1_subs-only\n",
            "{33: tensor([ 0.0144, -0.0011, -0.0693,  ..., -0.0230, -0.0688, -0.0119])}\n",
            "AncSARS-CoV-1_03-04_MAP\n",
            "{33: tensor([ 0.0095, -0.0220, -0.0646,  ..., -0.0386, -0.0439, -0.0491])}\n",
            "AncEurAf_altALL\n",
            "{33: tensor([ 0.0063, -0.0225, -0.0552,  ..., -0.0189, -0.0589, -0.0684])}\n",
            "AncSarbecovirus_MAP\n",
            "{33: tensor([ 0.0265, -0.0332, -0.0556,  ..., -0.0262, -0.0604, -0.0720])}\n",
            "AncSARS1c_MAP\n",
            "{33: tensor([ 0.0083, -0.0246, -0.0575,  ..., -0.0392, -0.0420, -0.0449])}\n",
            "AncSARS1a_MAP\n",
            "{33: tensor([-0.0324, -0.0559, -0.0650,  ..., -0.0284, -0.0549, -0.1151])}\n",
            "AncClade2_alt4_dels-only\n",
            "{33: tensor([ 0.0359, -0.0106, -0.0658,  ..., -0.0425, -0.0601, -0.0577])}\n",
            "AncSARS-CoV-1_02-03ML_MAP\n",
            "{33: tensor([ 0.0076, -0.0189, -0.0704,  ..., -0.0407, -0.0424, -0.0426])}\n",
            "AncSarbecovirus_altALL\n",
            "{33: tensor([ 0.0139, -0.0223, -0.0576,  ..., -0.0185, -0.0602, -0.0595])}\n",
            "AncSARS-CoV-1_02-03EML_MAP\n",
            "{33: tensor([ 0.0085, -0.0177, -0.0649,  ..., -0.0359, -0.0400, -0.0438])}\n",
            "AncAsia_tree1\n",
            "{33: tensor([ 0.0401, -0.0286, -0.0530,  ..., -0.0322, -0.0565, -0.0575])}\n",
            "AncSARS1a_tree1\n",
            "{33: tensor([ 0.0362, -0.0343, -0.0391,  ..., -0.0121, -0.0693, -0.0757])}\n",
            "AncClade2_MAP\n",
            "{33: tensor([ 0.0406, -0.0294, -0.0611,  ..., -0.0142, -0.0768, -0.0542])}\n",
            "AncAsia_altALL\n",
            "{33: tensor([ 0.0040, -0.0272, -0.0606,  ..., -0.0434, -0.0412, -0.0485])}\n",
            "AncSARS-CoV-1_altALL\n",
            "{33: tensor([ 0.0090, -0.0181, -0.0628,  ..., -0.0425, -0.0384, -0.0464])}\n",
            "AncSARS2c_MAP\n",
            "{33: tensor([ 0.0261, -0.0249, -0.0607,  ..., -0.0385, -0.0698, -0.0339])}\n",
            "AncClade2_tree2\n",
            "{33: tensor([ 0.0432, -0.0357, -0.0581,  ..., -0.0132, -0.0804, -0.0587])}\n",
            "AncSARS2a_altALL\n",
            "{33: tensor([ 0.0235, -0.0194, -0.0606,  ..., -0.0437, -0.0471, -0.0360])}\n",
            "AncSarbecovirus_alt1_ins117ins118\n",
            "{33: tensor([ 0.0166, -0.0325, -0.0547,  ..., -0.0244, -0.0662, -0.0696])}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Specify the folders containing the .pt files\n",
        "folder_paths = ['/content/coordoutputRBD.fasta','/content/coordoutputACE2.fasta',\n",
        "'/content/coordoutputRBDa.fasta']\n",
        "\n",
        "formatted_dict = {}\n",
        "\n",
        "# Iterate over each folder\n",
        "for folder_path in folder_paths:\n",
        "    # List all files in the folder with .pt extension\n",
        "    pt_files = [f for f in os.listdir(folder_path) if f.endswith('.pt')]\n",
        "\n",
        "    # Iterate over each .pt file in the current folder\n",
        "    for file_name in pt_files:\n",
        "        # Construct the full path to the file\n",
        "        file_path = os.path.join(folder_path, file_name)\n",
        "\n",
        "        # Load the model using torch.load\n",
        "        model_dict = torch.load(file_path, map_location=torch.device('cpu'))\n",
        "\n",
        "        # Extract label and tensor values\n",
        "        label = model_dict['label']\n",
        "        tensor_values = model_dict['mean_representations'][33].numpy()\n",
        "\n",
        "        # Convert tensor values to a list\n",
        "        tensor_values_list = tensor_values.tolist()\n",
        "\n",
        "        # Create a dictionary for the current file\n",
        "        file_dict = {label: tensor_values_list}\n",
        "\n",
        "        # Update the formatted_dict\n",
        "        formatted_dict.update(file_dict)\n",
        "\n",
        "\n",
        "print(formatted_dict)\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iNDOpSKEtgSP",
        "outputId": "e5fdf756-ec69-4f12-8845-cc07dcd68f9c"
      },
      "id": "iNDOpSKEtgSP",
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "IOPub data rate exceeded.\n",
            "The notebook server will temporarily stop sending output\n",
            "to the client in order to avoid crashing it.\n",
            "To change this limit, set the config variable\n",
            "`--NotebookApp.iopub_data_rate_limit`.\n",
            "\n",
            "Current values:\n",
            "NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
            "NotebookApp.rate_limit_window=3.0 (secs)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "\n",
        "csv_file_path = '/content/future learn data.csv'\n",
        "\n",
        "df = pd.read_csv(csv_file_path, header=None)\n",
        "name_value_mapping =formatted_dict\n",
        "\n",
        "# Replace all values in the DataFrame with the corresponding values from the mapping\n",
        "df = df.applymap(lambda x: name_value_mapping.get(x, x) if pd.notna(x) else x)\n",
        "\n",
        "# Save the modified DataFrame back to a CSV file\n",
        "df.to_csv('/content/future learn data numnum.csv', index=False)"
      ],
      "metadata": {
        "id": "fLyLgANjIDlA"
      },
      "id": "fLyLgANjIDlA",
      "execution_count": 17,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import csv\n",
        "\n",
        "input_csv_file = \"/content/future learn data numnum.csv\"\n",
        "output_csv_file = \"/content/future learn data num.csv\"\n",
        "\n",
        "\n",
        "with open(input_csv_file, 'r') as csv_in, open(output_csv_file, 'w', newline='') as csv_out:\n",
        "    reader = csv.reader(csv_in)\n",
        "    writer = csv.writer(csv_out)\n",
        "\n",
        "    for row in reader:\n",
        "        cleaned_row = [value.strip('[]') for value in row]\n",
        "        writer.writerow(cleaned_row)\n"
      ],
      "metadata": {
        "id": "kSCMAA2EOV0y"
      },
      "id": "kSCMAA2EOV0y",
      "execution_count": 18,
      "outputs": []
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.9.7"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
